stdlib/asm/collections/smt_new.masm

```diff
@@ -0,0 +1,161 @@
+        adv.push_mapval adv_push.1
+        # => [leaf_size, V_old, V, K, R]
+
+        # Leaf size will be a multiple of 8 (each kv-pair in a leaf is 8 elements)
+        dup eq.8
+        # => [is_single_kv_pair, leaf_size, V_old, V, K, R]
+        
+        if.true
+            #  Single kv-pair case
+
+            drop dropw adv_push.8
+            # => [K_in_leaf, V_old ,V, K, R]
+
+            # Check if the key stored in the leaf is the same
+            dupw.3 dupw.1 eqw
+            # => [K_in_leaf==K, K_in_leaf, V_old, V, K, R]
+
+            if.true
+                # Leaf stays a "single" variant
+
+                dropw
+                # => [V_old, V, K, R]
+
+                # Update advice map
+                movupw.2 movupw.2 adv.insert_hdword
+                # => [V, K, V_old, R]
+
+                # Prepare stack to update Merkle store
+                movupw.3 dup.11 push.DEPTH
+                # => [depth, K[3], R, V, K, V_old]
+
+                # Update Merkle store
+                mtree_set
+                # => [V_old_from_set, R_new, K, V_old_from_peek]
+
+                # Confirm that claimed `V_old` from `adv.push_smtpeek` is correct
+                movupw.3 dupw.1 assert_eqw
+                # => [V_old, R_new, K]
+
+                # Clean up stack for return
+                movupw.2 dropw
+                # => [V_old, R_new]
+            else
+                # Leaf becomes a Multiple kv-pair case
+                # TODO (fail for now)
+                push.1 assertz
+            end
+
+        else
+            # Multiple kv-pair case
+            # TODO (fail for now)
+            push.1 assertz
+        end
+    end
+end
+
+#! Input:  [K, R, ...]
+#! Output: [V, R, ...]
+export.get
+    # Prepare for `mtree_get`
+    # TODO: Confirm K[3] is indeed the MSB (i.e. are words pushed in reversed order on stack?)
+    dupw.1 dup.7 push.DEPTH
+    # => [depth, K[3], R, K, R]
+
+    # Retrieve node value from merkle store
+    mtree_get swapw dropw
+    # => [NV, K, R]
+
+    # Check if value is empty; if so, return empty value
+    padw eqw
+    # => [NV == 0, ZERO, V, K, R]
+
+    if.true
+        # Return empty value
+        dropw swapw dropw
+        # => [NV, R]
+    else
+        # Drop extra ZERO word
+        dropw
+        # => [NV, K, R]
+
+        # Get leaf pre-image from advice map. Push the leaf preimage size on the stack
+        adv.push_mapvaln adv_push.1
+        # => [leaf_size, NV, K, R]
+
+        # Leaf size will be a multiple of 8 (each kv-pair in a leaf is 8 elements)
+        dup eq.8
+        # => [is_single_kv_pair, leaf_size, NV, K, R]
+
+        if.true
+            # Single kv-pair case
+
+            # TODO: Does `adv_push` return `K, V` or `V, K`?
+            # As of currently, the order of inputs to `hmerge` below are
+            # inverted; `hmerge` should act on `[V, K]`.
+            # Push leaf pre-image on stack (single K-V pair)
+            drop adv_push.8
```

Comment:
This depends on how we initialize the advice provider. If the elements in the map are `HASH |-> [K, V]` then they will appear on the stack as `[V, K, ...]`.

I think we should assume that the advice provider is initialized in the correct (not inverted) order.